transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java

```diff
@@ -157,6 +158,65 @@ public final void read() {
         }
     }
 
+    private enum WriteStatus {
+        SET_OP_WRITE, // We have not written out the whole message and output channel is not writable
+        INCOMPLETE, // We have not written out the whole message after the spin count
+        DONE // we have written out the whole message
+    }
+
+    // returns whether we should setOpWrite
+    private WriteStatus doWriteFileRegion(ChannelOutboundBuffer in, FileRegion region,
+            int writeSpinCount) throws Exception {
+        if (region instanceof SliceableFileRegion) {
+            SliceableFileRegion sliceableRegion = (SliceableFileRegion) region;
+            if (!sliceableRegion.isTransferable()) {
+                return WriteStatus.DONE;
+            }
+            WriteStatus status = WriteStatus.INCOMPLETE;
+            long flushedAmount = 0;
+
+            for (int i = writeSpinCount - 1; i >= 0; i--) {
+                long localFlushedAmount = doWriteFileRegion(sliceableRegion);
+                if (localFlushedAmount == 0) {
+                    status = WriteStatus.SET_OP_WRITE;
+                    break;
+                }
+
+                flushedAmount += localFlushedAmount;
+                if (!sliceableRegion.isTransferable()) {
+                    status = WriteStatus.DONE;
+                    break;
+                }
+            }
+            in.progress(flushedAmount);
+            return status;
+        } else {
+            // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)